truffle/src/main/java/org/jruby/truffle/core/string/StringNodes.java

```diff
@@ -1355,6 +1365,307 @@ private int prevCharHead(Encoding enc, byte[]bytes, int p, int s, int end) {
+                        break;
+                    }
+                    else if (MBCLEN_CHARFOUND_P(ret)) {
+                        cr = CR_VALID;
+                        p += MBCLEN_CHARFOUND_LEN(ret);
+                    }
+                    else if (MBCLEN_INVALID_P(ret)) {
+                    /*
+                     * p1~p: valid ascii/multibyte chars
+                     * p ~e: invalid bytes + unknown bytes
+                     */
+                        int clen = enc.maxLength();
+                        if(buf == null){
+                            buf = RopeConstants.EMPTY_ASCII_8BIT_ROPE;
+                        }
+                        if (p > p1) {
+                               buf = makeConcatNode.executeMake(buf, RopeOperations.ropeFromByteList(new ByteList(pBytes, p1, p - p1)), enc);
+                        }
+
+                        if (e - p < clen){
+                            clen = e - p;
+                        }
+                        if (clen <= 2) {
+                            clen = 1;
+                        }
+                        else {
+                            int q = p;
+                            clen--;
+                            for (; clen > 1; clen--) {
+                                ret = enc.length(pBytes, q, q + clen);
+                                if (MBCLEN_NEEDMORE_P(ret)) break;
+                                if (MBCLEN_INVALID_P(ret)) continue;
+                            }
+                        }
+                        if (repBytes != null) {
+                            buf = makeConcatNode.executeMake(buf, RopeOperations.ropeFromByteList(new ByteList(repBytes, rep, replen)), enc);
+                            if (!rep7bit_p){
+                                cr = CR_VALID;
+                            }
+                        }
+                        else {
+                            repl = yield(frame, (DynamicObject)block, createString(new ByteList(pBytes, p, clen, enc, true)));
+                            repl = strCompatAndValidNode.call(frame, string, "str_compat_and_valid", repl, rubyEncoding);
+                            tainted |= isTaintedNode.executeIsTainted(repl);
+                            buf = makeConcatNode.executeMake(buf, rope((DynamicObject)repl), enc);
+                            if (rope((DynamicObject) repl).getCodeRange() == CR_VALID){
+                                cr = CR_VALID;
+                            }
+                        }
+                        p += clen;
+                        p1 = p;
+                        p = StringSupport.searchNonAscii(pBytes, p, e);
+                        if (p == -1) {
+                            p = e;
+                            break;
+                        }
+                    }
+                }
+                if (buf == null) {
+                    if (p == e) {
+//                        setCodeRange(cr);
```

Comment:
I'll need to look through this more, but I'd feel better if we have an assert that the code range is the same. Or maybe throw an exception if they're different. If they are different, there's a behavioral difference here and we may need to create a new rope with the correct code range and update the string's rope reference.